### ISSUE TITLE
Pytype fails to understand that typing.Union[<length-one-tuple>] is just <the one element of that length-one-tuple>

### ISSUE THREAD

**Issue by nathanielmanistaatgoogle** [google/pytype#622]
**DESCRIPTION:**
Pytype fails to understand some... admittedly pretty thorny and sophisticated code. But it fails in some weird ways.

**ENVIRONMENT:**
Python 3.8.3rc1
pytype 2020.6.26 or 2020.07.14

**STEPS TO REPRODUCE:**
Run pytype on
```python {.line-numbers}
import inspect
from typing import Any, Callable, Optional, Union

def _demonstration(behavior: Callable[..., Any]
) -> Optional[Callable[[str], Any]]:
    parameters = tuple(inspect.signature(behavior).parameters.items())
    for unused_parameter_name, parameter in parameters:
        if (getattr(parameter.annotation, '__origin__', None) is Union and
                type(None) in parameter.annotation.__args__):
            if 2 < len(parameter.annotation.__args__):
                raise ValueError('Complex sum types not yet supported!')
            effective_annotation: Callable[[str], Any] = Union[tuple(
                addend_type for addend_type in parameter.annotation.__args__
                if addend_type is not type(None))]
            #effective_annotation: Callable[[str], Any] = next(                                                                                                                                                                   
            #    iter(                                                                                                                                                                                                            
            #        addend_type                                                                                                                                                                                                  
            #        for addend_type in parameter.annotation.__args__                                                                                                                                                             
            #        if addend_type is not type(None)))                                                                                                                                                                           
            return effective_annotation
    return None
```
.

**EXPECTED RESULTS:**
pytype will not report any problems.

**OBSERVED RESULTS:**
pytype 2020.6.26 says
```
bad option in return type [bad-return-type]
           Expected: Optional[Callable[[str], Any]]
  Actually returned: Tuple[Callable[[str], Any], ...]
```
, which seems to suggest that pytype 2020.06.26 doesn't understand that the tuple constructed of the generator expression will be of length exactly one, and that Union, given such a tuple, will just evaluate to the tuple's one element. Separately, pytype 2020.07.14 says
```
Invalid type annotation '<instance of Tuple[Callable[[str], Any], ...]>' for effective_annotation [invalid-annotation]
  Not a type
```
which is very strange; the annotation really looks like a legitimate type...

**ADDITIONAL INFORMATION:**
By commenting-out the not-commented-out `effective_annotation` assignment statement and un-commenting-out the commented-out `effective_annotation` assignment statement one produces code that is acceptable to both pytype 2020.6.26 and 2020.07.14. Which makes the `Not a type` complaint for 2020.07.14 that much more strange, since the type annotation is exactly the same in both assignment statements.

**Comment by rchen152** [comment]
The specific errors you're getting are probably Python 3.8 weirdness, courtesy of our 3.8 support being incomplete. In 3.7, the (only) error I get with 2020.07.14 is:
```
File "foo.py", line 13, in _demonstration: Type annotation for effective_annotation does not match type of assignment [annotation-type-mismatch]
  Annotation: Callable[[str], Any]
  Assignment: tuple
```
which doesn't surprise me, given that I'm pretty sure we don't expect an annotation to ever be assigned as a value.

I'll make sure that pytype's behavior in 3.8 is fixed to match 3.7. As for the 3.7 issue...I'm not sure we really want to get into trying to analyze manipulation of type annotations.

**Comment by rchen152** [comment]
pytype's 3.8 behavior seems correct now (as in, the only error in 3.8 is the same one emitted in 3.7), so closing.
